scripts/gen_ldelf_hex.py

```diff
@@ -37,46 +37,65 @@ def round_up(n, m):
 
 def emit_load_segments(elffile, outf):
     load_size = 0
+    code_size = 0
     data_size = 0
-    next_rwseg_va = 0
+    load_segments = [s for s in elffile.iter_segments()
+                     if s['p_type'] == 'PT_LOAD']
+    load_segments.sort(key=lambda s: s['p_vaddr'])
+    prev_segment = None
+    pad_size = []
+    w_found = False
     n = 0
-    for segment in elffile.iter_segments():
-        if segment['p_type'] == 'PT_LOAD':
-            if n == 0:
-                if segment['p_flags'] != (P_FLAGS.PF_R | P_FLAGS.PF_X):
-                    print('Expected first load segment to be read/execute')
-                    sys.exit(1)
-                code_size = segment['p_filesz']
-            else:
-                if segment['p_flags'] != (P_FLAGS.PF_R | P_FLAGS.PF_W):
-                    print('Expected load segment to be read/write')
-                    sys.exit(1)
-                if next_rwseg_va and segment['p_vaddr'] != next_rwseg_va:
-                    print('Expected contiguous read/write segments')
-                    print(segment['p_vaddr'])
-                    print(next_rwseg_va)
-                    sys.exit(1)
-                data_size += segment['p_filesz']
-                next_rwseg_va = segment['p_vaddr'] + segment['p_filesz']
-            load_size += segment['p_filesz']
-            n = n + 1
-
+    # Check that load segments ordered by VA have the expected layout:
+    # read only first, then read-write. Compute padding.
```

Comment:
Compute padding at end of each segment, 0 if none is required.

Reply:
ok